oauth2-server/src/main/java/com/clouway/oauth2/CodeExchangeVerificationFlow.java

```diff
@@ -0,0 +1,43 @@
+package com.clouway.oauth2;
+
+import com.clouway.friendlyserve.Request;
+import com.clouway.friendlyserve.Response;
+import com.clouway.oauth2.authorization.Authorization;
+import com.clouway.oauth2.authorization.ClientAuthorizationRepository;
+import com.clouway.oauth2.client.Client;
+import com.clouway.oauth2.codechallenge.CodeVerifier;
+import com.google.common.base.Optional;
+
+/**
+ * @author Vasil Mitov <vasil.mitov@clouway.com>
+ */
+public class CodeExchangeVerificationFlow implements ClientActivity {
+  private final CodeVerifier codeVerifier;
+  private final ClientAuthorizationRepository clientAuthorizationRepository;
+  private final AuthorizedClientActivity clientActivity;
+
+  public CodeExchangeVerificationFlow(CodeVerifier codeVerifier, ClientAuthorizationRepository clientAuthorizationRepository, AuthorizedClientActivity clientActivity) {
+    this.codeVerifier = codeVerifier;
+    this.clientAuthorizationRepository = clientAuthorizationRepository;
+    this.clientActivity = clientActivity;
+  }
+
+  @Override
+  public Response execute(Client client, Request request, DateTime instant) {
+    String authCode = request.param("code");
+    String codeVerifierValue = request.param("code_verifier") == null ? "" : request.param("code_verifier");
+    Optional<Authorization> possibleAuthorization = clientAuthorizationRepository.findAuthorization(client, authCode, instant);
+
+    if (!possibleAuthorization.isPresent()) {
+      return OAuthError.invalidGrant();
+    }
+
+    Authorization authorization = possibleAuthorization.get();
```

Comment:
Loading of authorization and verification of code exchange need to be in separate implementations as they are not closely coupled.

// load authorization 
// verify code 
// issue token

Reply:
As i finished this change i think we could go one step further and add
//load authorization
//verify code
//load identity
//issue token 
That way every Activity will have 2 dependencies - the action it's doing and for the next activity to be executed. How does that sound ?

Reply:
The load identity could be the VerifiedCodeActivity

Reply:
Sounds good.

Reply:
Implemented and code has been rearranged